src/backend/opencl/cpu/cpu_sparse_blas.cpp

```diff
@@ -0,0 +1,530 @@
+                           (sparse_operation_t,
+                            scale_type<T>,
+                            const sparse_matrix_t,
+                            struct matrix_descr,
+                            cptr_type<T>,
+                            scale_type<T>,
+                            ptr_type<T>);
+
+template<typename T>
+using mm_func_def         = sparse_status_t (*)
+                           (sparse_operation_t,
+                            scale_type<T>,
+                            const sparse_matrix_t,
+                            struct matrix_descr,
+                            sparse_layout_t,
+                            cptr_type<T>,
+                            int, int,
+                            scale_type<T>,
+                            ptr_type<T>, int);
+
+#define SPARSE_FUNC_DEF( FUNC )                         \
+template<typename T> FUNC##_func_def<T> FUNC##_func();
+
+#define SPARSE_FUNC( FUNC, TYPE, PREFIX )               \
+  template<> FUNC##_func_def<TYPE> FUNC##_func<TYPE>()  \
+{ return &mkl_sparse_##PREFIX##_##FUNC; }
+
+SPARSE_FUNC_DEF( create_csr )
+SPARSE_FUNC(create_csr , float   , s)
+SPARSE_FUNC(create_csr , double  , d)
+SPARSE_FUNC(create_csr , cfloat  , c)
+SPARSE_FUNC(create_csr , cdouble , z)
+
+SPARSE_FUNC_DEF( mv )
+SPARSE_FUNC(mv , float   , s)
+SPARSE_FUNC(mv , double  , d)
+SPARSE_FUNC(mv , cfloat  , c)
+SPARSE_FUNC(mv , cdouble , z)
+
+SPARSE_FUNC_DEF( mm )
+SPARSE_FUNC(mm , float   , s)
+SPARSE_FUNC(mm , double  , d)
+SPARSE_FUNC(mm , cfloat  , c)
+SPARSE_FUNC(mm , cdouble , z)
+
+template<>
+const sp_cfloat getScaleValue<const sp_cfloat, cfloat>(cfloat val)
+{
+    sp_cfloat ret;
+    ret.real = val.s[0];
+    ret.imag = val.s[1];
+    return ret;
+}
+
+template<>
+const sp_cdouble getScaleValue<const sp_cdouble, cdouble>(cdouble val)
+{
+    sp_cdouble ret;
+    ret.real = val.s[0];
+    ret.imag = val.s[1];
+    return ret;
+}
+
+#else   // USE_MKL
+
+// From mkl_spblas.h
+typedef enum
+{
+    SPARSE_OPERATION_NON_TRANSPOSE          = 10,
+    SPARSE_OPERATION_TRANSPOSE              = 11,
+    SPARSE_OPERATION_CONJUGATE_TRANSPOSE    = 12,
+} sparse_operation_t;
+
+#endif  // USE_MKL
+
+sparse_operation_t
+toSparseTranspose(af_mat_prop opt)
+{
+    sparse_operation_t out = SPARSE_OPERATION_NON_TRANSPOSE;
+    switch(opt) {
+        case AF_MAT_NONE        : out = SPARSE_OPERATION_NON_TRANSPOSE;         break;
+        case AF_MAT_TRANS       : out = SPARSE_OPERATION_TRANSPOSE;             break;
+        case AF_MAT_CTRANS      : out = SPARSE_OPERATION_CONJUGATE_TRANSPOSE;   break;
+        default                 : AF_ERROR("INVALID af_mat_prop", AF_ERR_ARG);
+    }
+    return out;
+}
+
+template<typename T, int value>
+scale_type<T> getScale()
+{
+    static T val = scalar<T>(value);
+    return getScaleValue<scale_type<T>, T>(val);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+#ifdef USE_MKL // Implementation using MKL
```

Comment:
I understand. But you can wrap the two backends into a similar interface and call it from a single function instead of making two versions of the same thing that share the same initial logic.